data/src/hash.rs

```diff
@@ -35,52 +35,75 @@ pub enum HashError {
 }
 
 /// Size of digest produced by the underlying hash function
-pub const DIGEST_SIZE: usize = 32;
+const DIGEST_SIZE: usize = 32;
+
+/// [struct@HashDigest] is a wrapper class around [u8; 32] because
+/// we would like to have a unified way of using hashing.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Encode, Decode, Hash, PartialOrd, Ord)]
+pub struct HashDigest([u8; DIGEST_SIZE]);
+
+impl HashDigest {
+    /// Returns the number of bytes a [`HashDigest`] is made up of.
+    pub const fn len() -> usize {
+        DIGEST_SIZE
+    }
+}
+
+impl AsRef<[u8]> for HashDigest {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+}
+
+impl From<blake3::Hash> for HashDigest {
+    fn from(value: blake3::Hash) -> Self {
+        Self(value.into())
+    }
+}
+
+impl From<[u8; DIGEST_SIZE]> for HashDigest {
+    fn from(value: [u8; DIGEST_SIZE]) -> Self {
+        Self(value)
+    }
+}
+
+impl From<HashDigest> for [u8; DIGEST_SIZE] {
+    fn from(value: HashDigest) -> Self {
+        value.0
+    }
+}
 
 /// A value of type [struct@Hash] indicates that the enclosed array is a digest
 /// produced by a preset hash function, currently BLAKE2b. It can be obtained
 /// by either hashing data directly or after hashing by converting from
 /// a suitably sized byte slice or vector.
-#[derive(
-    Clone,
-    Copy,
-    PartialEq,
-    Eq,
-    Encode,
-    Decode,
-    Hash,
-    PartialOrd,
-    Ord,
-    derive_more::From,
-    derive_more::Debug,
-)]
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, Hash, PartialOrd, Ord, derive_more::Debug)]
 #[debug("{}", self)]
 pub struct Hash {
-    digest: [u8; DIGEST_SIZE],
+    digest: HashDigest,
 }
 
 impl Hash {
-    /// Hash a slice of bytes
-    pub fn blake3_hash_bytes(bytes: &[u8]) -> Self {
+    /// Hashes a byte slice into a [`struct@Hash`] object.
+    pub fn hash_bytes(bytes: &[u8]) -> Self {
         let digest = blake3::hash(bytes).into();
         Hash { digest }
     }
 
-    /// Get the hash of a value that can be serialised by hashing its serialisation
-    pub fn blake3_hash<T: Encode>(data: T) -> Result<Self, EncodeError> {
+    /// Creates a [`struct@Hash`] object from something that implements the
+    /// [`bincode::enc::Encode`] trait.
+    pub fn hash_encodable<T: Encode>(data: T) -> Result<Self, EncodeError> {
```

Comment:
```suggestion
    pub fn hash<T: Encode>(data: T) -> Result<Self, EncodeError> {
```
Same here, the `_encodable` is clear from the function's signature.